### PR TITLE
feat: fix prorating credit corrections

### PR DESCRIPTION
**Pull request by turip**
## Overview

Extends credit-then-invoice subscription sync coverage and the flat-fee CTI behavior needed by those scenarios. 

## Code Changes

- Add `sync_credittheninvoice_test.go` as the CTI subscription-sync suite, with converted scenarios for happy path provisioning, prorating, gathering/draft/issued invoice flows, aligned invoicing/cancellation, one-time fees, rate-card tax propagation, instant billing, usage-based updates, and ledger balance expectations.
- Keep the remaining copied CTI scenarios present but skipped where they still need conversion, and keep `TestDiscountSynchronization` skipped for the known flat-fee CTI discount allocation issue.
- Add shared subscription-sync charge assertion helpers on `SuiteBase`, including exact charge count matching, expected charge status, settlement mode, periods, invoice-at where applicable, tax config, and subscription linkage.
- Add flat-fee CTI zero-amount handling: zero-amount charges are still tracked as charges, but they do not materialize billable gathering invoice lines.
- Add flat-fee CTI mutable standard-line reconciliation so charge runs, credit allocations/corrections, detailed lines, and totals are reconciled when a mutable draft line is rebuilt from an updated intent.
- Add the `CreditNotesSupportedByLineUpdater` default feature flag for flat-fee CTI immutable-proration behavior. With the default unsupported path, immutable prorated lines record the unsupported-credit-note warning without creating replacement billable work for the already-invoiced period.
- Add charges service tests for zero-amount CTI flat fees and immutable proration with credit notes both available and unavailable.
- Move `TestUncollectableCollection` from subscription sync tests into `test/billing/collection_test.go`.

## Notes for Reviewer

- Manual edit/delete/ignore sync scenarios are intentionally absent from the CTI suite because charge-managed items cannot be manually edited through invoice-line sync.
- The original generic rate-card tax testcase is represented as CTI-specific `TestRateCardTaxSyncFlatFee` and `TestRateCardTaxSyncUsageBased` cases.
- `TestDiscountSynchronization` remains skipped because flat-fee CTI currently allocates credits before percentage discounts are applied, which causes the discounted line to fail with credits-not-consumed validation.

## Validation

- `go test -count=1 -tags=dynamic ./openmeter/billing/worker/subscriptionsync/service -run '^TestCreditThenInvoiceScenarios$'`
- `make lint-go-fast`

## Summary by CodeRabbit

* **New Features**
  * Toggle to control credit-note support for flat-fee billing behavior.

* **Bug Fixes**
  * Zero-amount charges no longer create gathering or invoice lines.

* **Improvements**
  * Targeted credit reconciliation with timestamped allocation/reconciliation.
  * Improved reconciliation of invoice lines to updated intents; clearer immutable vs mutable update handling.

* **Tests**
  * New and expanded tests for zero-amount, immutable/mutable proration, reconciliation flows.

* **Documentation**
  * Added guidance on helper usage and a utility library usage note.

[Review Change Stack](https://app.coderabbit.ai/change-stack/openmeterio/openmeter/pull/4363)

### CODE REVIEW

File: openmeter/billing/charges/flatfee/service/service.go [hunk 1]
```diff
@@ -67,24 +68,30 @@ func New(config Config) (flatfee.Service, error) {
...
 
 func (s *service) GetLineEngine() billing.LineEngine {
 	return &LineEngine{
 		service: s,
 	}
 }
+
+func (s *service) SetCreditNotesSupportedByLineUpdater(supported bool) {
+	s.creditNotesSupported = supported
+}
```
[the remark this answers is not in the document]
Comment (turip) [reply openmeter/billing/charges/flatfee/service/service.go]: This is internal only, safe to ignore.

File: openmeter/billing/charges/flatfee/service/realizations/correct.go [hunk 2]
```diff
@@ -5,13 +5,146 @@ import (
...
+type ReconcileCreditRealizationsInput struct {
+	Charge             flatfee.Charge
+	Run                flatfee.RealizationRun
+	AllocateAt         time.Time
+	TargetAmount       alpacadecimal.Decimal
+	CurrencyCalculator currencyx.Calculator
+}
+
+func (i ReconcileCreditRealizationsInput) Validate() error {
+	if err := i.Charge.Validate(); err != nil {
```
Comment (tothandras) [inline openmeter/billing/charges/flatfee/service/realizations/correct.go:32]: Why don't we use `errs` pattern with generic validation error here?

**Review by tothandras — CHANGES_REQUESTED** [review]

File: openmeter/billing/charges/flatfee/service/realizations/correct.go (hunk 2 above)
Comment (turip) [reply openmeter/billing/charges/flatfee/service/realizations/correct.go:32]: usually I just do that on service boundary, this is too deep to have meaningful validation erros, but so be it ;-)

**Review by tothandras — APPROVED** [review]
